### PR TITLE
Fix summarise trips without trips

### PR DESCRIPTION
**Pull request by allenmichael099**
@asiripanich Shankari and I noticed today that there was an error with some of the trip summary columns. All participants had the same value. Also, the participation period plot was uniform. After looking into it, I realized that the mistake was in how I implemented summarise_trips_without_trips. It was using the min and max dates for all trips, rather than the min and max for each user. 

It seems to be fixed but perhaps it should be more carefully tested. I threw together the beginnings of a test that for it in test-utils_tidy_data, and the columns are mostly the same. One version seems to round the timestamps to the nearest day while the other does not. n_active_days might appear in a different order.

Here are relevant screenshots of the issue  (see last trip date and number of days since app install):
Wrong_data_upload
Wrong_particip_pd

And after the fix:
Fixed_data_upload
Fixed_Particip_Pd

### CODE REVIEW

**Comment by asiripanich** [comment]
@allenmichael099 just wondering why did you comment that test out? c821219

**Comment by allenmichael099** [comment]
@asiripanich I figured it wasn't ready yet, and I didn't know what expectation to set.

**Comment by asiripanich** [comment]
@allenmichael099 Could you please submit another PR with unit tests, after Aug is fine. Thanks!

File: R/utils_tidy_data.R [hunk 1]
```diff
@@ -192,12 +198,22 @@ summarise_trips_without_trips <- function(participants, cons) {
...
+  
+  # Count the number of trips per user
   n_trips <- count_total_trips(cons)
   summ_trips <- merge(n_trips, summ_trips, by = "user_id")
 
   message("merging trip summaries with participants")
-  merge(participants, summ_trips, by = "user_id", all.x = TRUE)
+  merge(participants, summ_trips, by = "user_id", all.x = TRUE) %>% 
+    merge(., unconfirmed_summ, by="user_id", all.x = TRUE) %>%
+    .[is.na(unconfirmed), unconfirmed := 0]
```
Comment (shankari) [inline R/utils_tidy_data.R:213]: Note to self: checked that this is the same as in my `master_specific_customization` branch

File: R/utils_tidy_data.R [hunk 2]
```diff
@@ -192,12 +198,22 @@ summarise_trips_without_trips <- function(participants, cons) {
       last_trip_local_datetime = format(max(lubridate::as_datetime(end_local_time)), usetz = FALSE)
     ), by = user_id] %>%
     .[, n_days := round(as.numeric(difftime(last_trip_datetime, first_trip_datetime, units = "days")), 1)]
-
+  
+  # Add the 'unconfirmed' column.
+  unconfirmed_summ <- trip_query %>%
+    .[is.na(trip_query[[confirmed_user_input_column]]), .(unconfirmed = .N), by = user_id]
+
```
Comment (shankari) [inline R/utils_tidy_data.R:205]: note to self, this makes sense and is also similar to my original code.

File: R/utils_tidy_data.R [hunk 3]
```diff
@@ -162,27 +162,33 @@ tidy_cleaned_trips_by_timestamp <- function(df) {
...
     normalise_uuid()
-  first_trip_local_datetime <- format(min(lubridate::as_datetime(start_local_time)), usetz = FALSE)
-
+  
+  # Generate intermediate columns to use for summarizing trips.
+  helper_trip_cols <- trip_query %>%
+    .[, .(
+      user_id = user_id,
+      start_fmt_time = lubridate::as_datetime(trip_query$data.start_fmt_time), # convert trip start datetimes to UTC
+      end_fmt_time = lubridate::as_datetime(trip_query$data.end_fmt_time) # convert trip end datetimes to UTC
```
Comment (shankari) [inline R/utils_tidy_data.R:175]: note to self: this is similar to the original "before" code.

File: R/utils_query_database.R [hunk 4]
```diff
@@ -333,16 +333,19 @@ count_total_trips <- function(cons) {
...
-                                  "data.end_local_dt":true,
-                                  "data.end_fmt_time":true,
-                                  "user_id":true, "_id":false}'
+    fields = sprintf('{"data.start_local_dt":true,
+              "data.start_fmt_time":true,
+              "data.end_local_dt":true,
+              "data.end_fmt_time":true,
+              "%s":true,
+              "user_id":true, 
+              "_id":false}', confirmed_user_input_column)
```
Comment (shankari) [inline R/utils_query_database.R:346]: note to self: these fields seem fine and the `sprintf` makes sense. Thanks R for using standard C function name.

File: R/utils_tidy_data.R [hunk 5]
```diff
@@ -162,27 +162,33 @@ tidy_cleaned_trips_by_timestamp <- function(df) {
...
+    )] %>%
+    # Now add columns generated from start/end_fmt_time
+    .[, ":="
+    (
+      start_local_time = purrr::map2(start_fmt_time, trip_query$data.start_local_dt.timezone, ~ format(.x, tz = .y, usetz = TRUE)) %>%
+        as.character(), # format returns a list, but lubridate::as_datetime needs it as a character string
+      end_local_time = purrr::map2(end_fmt_time, trip_query$data.end_local_dt.timezone, ~ format(.x, tz = .y, usetz = TRUE)) %>%
+        as.character()
```
Comment (shankari) [inline R/utils_tidy_data.R]: Ditto

**Review by shankari — CHANGES_REQUESTED** [review]
The code changes all make sense to me. Why is the test failing?

**Comment by asiripanich** [comment]
@allenmichael099  Now that we have `summarise_trips_without_trips()`, do we still need `summarise_trips()`

**Comment by allenmichael099** [comment]
@asiripanich I kept summarise_trips to test whether the output of summarise_trips_without_trips is different than it should be. It might still be useful to have summarise_trips to compare with the output from summarise_trips_without_trips after any future changes to summarise_trips_without_trips.

**Comment by asiripanich** [comment]
> 
> 
> @asiripanich I kept summarise_trips to test whether the output of summarise_trips_without_trips is different than it should be. It might still be useful to have summarise_trips to compare with the output from summarise_trips_without_trips after any future changes to summarise_trips_without_trips.

That is what I thought you were trying to do. Instead of that, we can use `testthat` to take a snapshot of the output and test against that. See https://github.com/asiripanich/emdash/blob/master/tests/testthat/_snaps/utils_tidy_data.md.

**Pull request merged**